File: worker/worker.py
import sys, json, importlib, traceback

# Stray prints from imported libraries must not corrupt the protocol stream.
proto = sys.stdout
sys.stdout = sys.stderr

objs = {}
next_id = 0


def send(msg):
    proto.write(json.dumps(msg) + "\n")
    proto.flush()


def read_msg():
    while True:
        line = sys.stdin.readline()
        if not line:
            sys.exit(0)
        if line.strip():
            return json.loads(line)


def to_wire(v):
    global next_id
    if hasattr(v, "item"):  # numpy-style scalars -> plain numbers
        try:
            v = v.item()
        except Exception:
            pass
    try:
        json.dumps(v, allow_nan=False)
        return {"$": "data", "v": v}
    except (TypeError, ValueError, OverflowError):
        next_id += 1
        objs[next_id] = v
        return {"$": "ref", "id": next_id, "repr": repr(v).split("\n")[0][:80]}


def from_wire(w):
    if isinstance(w, dict):
        t = w.get("$")
        if t == "ref":
            return objs[w["id"]]
        if t == "fn":
            return make_callback(w["id"])
        return {k: from_wire(v) for k, v in w.items()}
    if isinstance(w, list):
        return [from_wire(x) for x in w]
    return w


def make_callback(fn_id):
    def cb(*args, **kwargs):
        if kwargs:
            raise TypeError("seam functions don't take keyword arguments")
        send({"cb": True, "fn": fn_id, "args": [to_wire(a) for a in args]})
        while True:
            msg = read_msg()
            if msg.get("cbr"):
                if msg.get("ok"):
                    return from_wire(msg["value"])
                raise RuntimeError("seam callback failed: " + msg.get("error", "unknown"))
            # the host re-entered us while evaluating the callback
            handle_request(msg)

    return cb


def handle_request(req):
    try:
        op = req["op"]
        if op == "import":
            res = importlib.import_module(req["name"])
        elif op == "getattr":
            res = getattr(from_wire(req["obj"]), req["name"])
        elif op == "index":
            res = from_wire(req["obj"])[from_wire(req["key"])]
        elif op == "call":
            res = from_wire(req["obj"])(
                *[from_wire(a) for a in req.get("args", [])],
                **{k: from_wire(v) for k, v in req.get("kwargs", {}).items()},
            )
        elif op == "str":
            res = str(from_wire(req["obj"]))
        elif op == "release":
            objs.pop(req["ref"], None)
            res = None
        else:
            raise ValueError(f"unknown op: {op}")
        out = {"id": req["id"], "ok": True, "value": to_wire(res)}
    except Exception as e:
        out = {
            "id": req["id"],
            "ok": False,
            "error": f"{type(e).__name__}: {e}",
            "trace": traceback.format_exc(),
        }
    send(out)


while True:
    handle_request(read_msg())

File: worker/worker.js
// seam node worker — NDJSON request/response on stdout; console goes to stderr
const readline = require('readline');
const util = require('util');
const path = require('path');
const { createRequire } = require('module');
const { pathToFileURL } = require('url');

// Resolve bare specifiers against the script's cwd, so node_modules just works.
const cwdRequire = createRequire(path.join(process.cwd(), '__seam__.js'));

// Stray console output from libraries must not corrupt the protocol.
const proto = process.stdout;
console.log = (...a) => console.error(...a);
console.info = (...a) => console.error(...a);
console.warn = (...a) => console.error(...a);

const objs = new Map();
let nextId = 0;

// Data = JSON-shaped values only. Class instances, functions, Maps, NaN,
// BigInt etc. stay here as refs — JSON.stringify would silently mangle them.
function isData(v) {
  if (v === null || typeof v === 'boolean' || typeof v === 'string') return true;
  if (typeof v === 'number') return Number.isFinite(v);
  if (Array.isArray(v)) return v.every(isData);
  if (typeof v === 'object') {
    const p = Object.getPrototypeOf(v);
    return (p === Object.prototype || p === null) && Object.values(v).every(isData);
  }
  return false;
}

function toWire(v) {
  if (v === undefined) return { $: 'data', v: null };
  if (isData(v)) return { $: 'data', v };
  nextId += 1;
  objs.set(nextId, v);
  const repr = util.inspect(v, { depth: 1 }).split('\n')[0].slice(0, 80);
  return { $: 'ref', id: nextId, repr };
}

function fromWire(w) {
  if (Array.isArray(w)) return w.map(fromWire);
  if (w && typeof w === 'object') {
    if (w.$ === 'ref') return objs.get(w.id);
    if (w.$ === 'fn') return makeCallback(w.id);
    const o = {};
    for (const [k, v] of Object.entries(w)) o[k] = fromWire(v);
    return o;
  }
  return w;
}

// Seam functions arrive as handles. Node can't block on stdin, so invoking
// one returns a promise; replies resolve LIFO (nesting is strict).
const pendingCbs = [];

function makeCallback(fnId) {
  return (...args) =>
    new Promise((resolve, reject) => {
      pendingCbs.push({ resolve, reject });
      proto.write(JSON.stringify({ cb: true, fn: fnId, args: args.map(toWire) }) + '\n');
    });
}

// ESM namespaces hide the useful thing behind .default — unwrap it.
function preferDefault(mod) {
  if (mod && (mod[Symbol.toStringTag] === 'Module' || mod.__esModule) && 'default' in mod) {
    return mod.default;
  }
  return mod;
}

async function importModule(name) {
  try {
    return preferDefault(cwdRequire(name));
  } catch (e) {
    let spec = name;
    try { spec = pathToFileURL(cwdRequire.resolve(name)).href; } catch {}
    return preferDefault(await import(spec));
  }
}

async function handle(req) {
  switch (req.op) {
    case 'import':
      return await importModule(req.name);
    case 'getattr': {
      const obj = fromWire(req.obj);
      const res = obj[req.name];
      return typeof res === 'function' ? res.bind(obj) : res;
    }
    case 'index':
      return fromWire(req.obj)[fromWire(req.key)];
    case 'call': {
      const fn = fromWire(req.obj);
      if (typeof fn !== 'function') throw new TypeError('value is not a function');
      const args = (req.args || []).map(fromWire);
      const kw = req.kwargs || {};
      if (Object.keys(kw).length > 0) args.push(fromWire(kw)); // kwargs -> options object
      return await fn(...args);
    }
    case 'str':
      return util.inspect(fromWire(req.obj), { depth: 2 });
    case 'release':
      objs.delete(req.ref);
      return null;
    default:
      throw new Error('unknown op: ' + req.op);
  }
}

const rl = readline.createInterface({ input: process.stdin, terminal: false });
rl.on('line', async (line) => {
  if (!line.trim()) return;
  const req = JSON.parse(line);
  if (req.cbr) {
    const p = pendingCbs.pop();
    if (p) {
      req.ok ? p.resolve(fromWire(req.value)) : p.reject(new Error('seam callback failed: ' + req.error));
    }
    return;
  }
  let res;
  try {
    res = { id: req.id, ok: true, value: toWire(await handle(req)) };
  } catch (e) {
    res = {
      id: req.id,
      ok: false,
      error: e instanceof Error ? `${e.name}: ${e.message}` : String(e),
      trace: (e instanceof Error && e.stack) || '',
    };
  }
  proto.write(JSON.stringify(res) + '\n');
});
